python/dolfinx/fem/petsc.py

```diff
@@ -634,86 +559,147 @@ def _assemble_matrix_block_mat(
+    x0: typing.Optional[Iterable[PETSc.Vec]] = None,
     alpha: float = 1,
-    constants=None,
+    constants: typing.Optional[
+        typing.Union[Iterable[np.ndarray], Iterable[Iterable[np.ndarray]]]
+    ] = None,
     coeffs=None,
 ) -> None:
-    """Apply the function :func:`dolfinx.fem.apply_lifting` to a PETSc Vector."""
-    with contextlib.ExitStack() as stack:
-        x0 = [stack.enter_context(x.localForm()) for x in x0]
-        x0_r = [x.array_r for x in x0]
-        b_local = stack.enter_context(b.localForm())
-        _assemble.apply_lifting(b_local.array_w, a, bcs, x0_r, alpha, constants, coeffs)
+    """Modify the right-hand side PETSc vector ``b`` to account for
+    constraints (Dirichlet boundary conitions).
 
+    See :func:`dolfinx.fem.apply_lifting` for a mathematical
+    descriptions of the lifting operation.
+
+    Args:
+        b: Vector to modify in-place.
+        a: List of bilinear forms. If ``b`` is not blocked or a nest,
+            then ``a`` is a 1D sequence. If ``b`` is blocked or a nest,
+            then ``a`` is  a 2D array of forms, with the ``a[i]`` forms
+            used to modify the block/nest vector ``b[i]``.
+        bcs: Boundary conditions to use to modify ``b`` (see
+            :func:`dolfinx.fem.apply_lifting`). The boundary conditions
+            in ``bcs[j]`` are associated with the forms in the ``j``th
+            colulmn of ``a``.
+        x0: Vector to use in modify ``b`` (see
+            :func:`dolfinx.fem.apply_lifting`). Treated as zero if
+            ``None``.
+        alpha: Scalar parameter in lifting (see
+            :func:`dolfinx.fem.apply_lifting`).
+        constants: Packed constant data appearing in the forms ``a``. If
+            ``None``, the constant data will be packed by the function.
+        coeffs: Packed coefficient data appearing in the forms ``a``. If
+            ``None``, the coefficient data will be packed by the
+            function.
+
+    Note:
+        Ghost contributions are not accumulated (not sent to owner).
+        Caller is responsible for reverse-scatter to update the ghosts.
+
+    Note:
+        Boundary condition values are *not* set in ``b`` by this
+        function. Use :func:`dolfinx.fem.DirichletBC.set` to set values
+        in ``b``.
+    """
+    if b.getType() == PETSc.Vec.Type.NEST:
+        x0 = [] if x0 is None else x0.getNestSubVecs()
+        constants = [pack_constants(forms) for forms in a] if constants is None else constants
+        coeffs = [pack_coefficients(forms) for forms in a] if coeffs is None else coeffs
+        for b_sub, a_sub, const, coeff in zip(b.getNestSubVecs(), a, constants, coeffs):
+            const_ = list(
+                map(lambda x: np.array([], dtype=PETSc.ScalarType) if x is None else x, const)
+            )
+            apply_lifting(b_sub, a_sub, bcs, x0, alpha, const_, coeff)
+    else:
+        with contextlib.ExitStack() as stack:
```

Comment:
This is where I think it would be natural to extract he appropriate bcs from the list of BCs.